WIP: basic correctness test

derekk-nm opened the pull request:
Introducing an end-to-end test case that verifies basic correctness of the vllm engine by comparing the tokens output by the vllm OpenAI server with tokens generated by the HuggingFace model created by `AutoModelForCausalLM.from_pretrained()`.
This Work In Progress PR is intended to verify the approach used to reconcile the different logprobs output generated by the server (keyed by tokens) vs that generated by the HF model (keyed by token id). It uses duplicate code in the `tests.conftest.HfRunnerNM.generate_greedy_logprobs_nm_use_tokens()` method, replacing a token id with the actual token. As long as this approach is approved, I'll come up with some code refactoring to reduce this, and other code duplication used in other fixtures in `conftest.py`.
The use of the new `VllmServer` class is also an open question as we are still determining the actual infrastructure that will be used to execute this test. [credit to Domenic Barbuzzi for the VllmServer code]

derekk-nm commented on tests/models/compare_utils.py:
this edit should have been a copy of the existing code.  clearly, these changes will break other tests using this method.

dbarbuzzi replied:
I think it would make sense to make a copy somewhere inside the `neuralmagic` folder for this usage.

dbarbuzzi commented:
Thanks @derekk-nm, I think the new test is looking good!

>     # now repeat using two gpus
>     # specifically doing it here, rather than as a pytest param,
>     # to avoid repeating the huggingface inference and data collection

If this test isn’t getting forked through `pytest-forked` or similar, I can work on a cached helper function to generate the referenced outputs, so using 1 vs. 2 GPUs can be separate tests while still only incurring the cost of the HF output generation once.

derekk-nm commented:
Another open question to resolve w/ this code is the handling of spaces in tokens.  

here's an image from my debugger showing that huggingface generated text does not have spaces in the tokens, but the openai server's response does.
(I did try using clean_up_tokenization_spaces=False on the huggingface tokenizer's decode method when I retrieved the token's string, but that didn't help).
<img width="603" alt="generated tokens and leading spaces diff" src="https://github.com/neuralmagic/nm-vllm/assets/106621615/f0d8a62e-62a0-4e80-aefb-41731b941afa">

Also, another image of an "empty" response from one of the other prompts that includes that </s> marker.  we'll have to deal w/ this too.
<img width="793" alt="Screenshot 2024-04-17 at 7 31 57 AM" src="https://github.com/neuralmagic/nm-vllm/assets/106621615/76748b32-0c14-4a27-b854-f17b98c09a6e">

derekk-nm commented:
> Thanks @derekk-nm, I think the new test is looking good!
> 
> > ```
> > # now repeat using two gpus
> > # specifically doing it here, rather than as a pytest param,
> > # to avoid repeating the huggingface inference and data collection
> > ```
> 
> If this test isn’t getting forked through `pytest-forked` or similar, I can work on a cached helper function to generate the referenced outputs, so using 1 vs. 2 GPUs can be separate tests while still only incurring the cost of the HF output generation once.

As far as I can tell, the tests **are** currently run with `--forked`, at least the existing tests.  I don't know if we'll end up with some different architecture to run end-to-end tests.

dbarbuzzi commented:
> As far as I can tell, the tests are currently run with --forked, at least the existing tests. I don't know if we'll end up with some different architecture to run end-to-end tests.

In that case, maybe we can review tests so only tests that require being forked are (by using the pytest marker instead of the CLI flag) or otherwise splitting into multiple runs to be able to not fork this one. Overall, I don’t think such an enhancement is required for the initial implementation.

dbarbuzzi commented on neuralmagic/tests/test_basic_correctness.py:
```diff
@@ -0,0 +1,124 @@
+    # now repeat using two gpus
+    # specifically doing it here, rather than as a pytest param,
+    # to avoid repeating the huggingface inference and data collection
+    api_server_args["--tensor-parallel-size"] = 2
```
With our current server helper implementation, this needs to be a string (the args dict currently must be all strings for names and values). The helper class should actually raise here due to this value not being a string, if it’s not then something is going wrong in the helper…

derekk-nm replied:
the test execution hasn't gotten this far yet because of the issue w/ spaces in the tokens.  I'll update the value.

dbarbuzzi replied:
That makes sense, but even in the initial definition of ``, there are non-string values which should be causing the `VllmServer` class to raise a `ValueError` while processing the dict.

dhuangnm replied:
Should this --tensor-parallel-size be treated similarly as pytest.mark.parametrize for other arguments e.g. model etc? There is no need to repeat the test code here just for --tensor-parallel-size=2. Also it can be specified to different value than 2 based on the gpu type.

derekk-nm replied:
I made an execution error, so I didn't actually test this.  I'll pass in a string.

derekk-nm replied:
@dhuangnm , the problem w/ passing `tensor-parallel-size` as a param is that it would result in the execution of the  code that collects the huggingface data twice.  this is expensive in execution time.

Maybe we could create a session scoped fixture that stores the results the first time, and retrieves the results afterward?

As far as wanting to set the `tensor-parallel-size` arg to a different type based on gpu type, if there's a way to determine that programmatically, we could create a fixture for that.

dbarbuzzi replied:
> Maybe we could create a session scoped fixture that stores the results the first time, and retrieves the results afterward?

This also would not survive `pytest-forked`

derekk-nm replied:
There is a potential solution in the [pypi doc for the pytest-xdist-forked](https://pypi.org/project/pytest-xdist-forked/) package.  see the section **Making session-scoped fixtures execute only once**.  They basically suggest "using a lock file for inter-process communication".  (I'm assuming that it should work w/ the plain pytest-xdist package, but I don't know that.)

dhuangnm replied:
I think we need to think of a better way for the parameters that can be passed to the server which we may want to test but are not needed to get the HF results. It's not ideal to repeat test code for each parameter.

robertgshaw2-neuralmagic replied:
`torch.cuda.device_count` may work for our purposes

derekk-nm replied:
This conversation is basically moot now that we've decided to have the server and clients run in the same test.

dbarbuzzi commented on neuralmagic/tests/test_basic_correctness.py:
```diff
@@ -0,0 +1,124 @@
+    logger = logging.Logger("vllm_server")
+    api_server_args = {
+        "--model": model,
+        "--max-model-len": max_model_len,
```
Similar to my other comment, this should be causing failure in the helper class due to `max_model_len` not being a string…

Which leads to two comments:
* If this `ValueError` is not being raised, something is going wrong within the helper
* If we want ‘natural’ data types for this arg dict (e.g. ints, bools) then that can be taken care of with some internal refactoring of `VllmServer` (won’t affect external usage other than allowing non-string values).

dbarbuzzi commented:
@derekk-nm I added a couple of comments about the arg list. Per a previous discussion, I left the initial implementation in where the `args` dict passed to the `VllmServer` constructor expects all keys and values to be strings, and _should_ actually be throwing errors in your tests due to some of the values in the dicts being ints, floats, and `None`.

So, summarizing some comments:
* If the test and server helper are working as-is, something is going wrong with input validation
    * Actually, going even further, even if the built-in error-checking is not passing, the "command" list passed to something like `Popen` basically requires strings (technically also bytes or Pathlike objects) so no matter what, passing ints or `None` as values should not actually work…
* All that said, we can certainly amend `VllmServer` so that this args dict can have non-string values, that was a design question and simply requires some refactoring of the logic within the `_args_to_list` private method (no other external usage changes).

dbarbuzzi commented:
> Should this --tensor-parallel-size be treated similarly as pytest.mark.parametrize for other arguments e.g. model etc? There is no need to repeat the test code here just for --tensor-parallel-size=2. Also it can be specified to different value than 2 based on the gpu type.

I believe per the comments that this is currently specifically part of the same test so that the hugging-face outputs only need to be generated once (as they are needed for a baseline comparison after both server portions).

robertgshaw2-neuralmagic commented on neuralmagic/tests/test_basic_correctness.py:
```diff
@@ -0,0 +1,124 @@
+        api_server_args["--sparsity"] = sparsity
+
+    with VllmServer(api_server_args, logger):
+        completion = await client.completions.create(model=model,
```
This is doing something different than the tests that we previously created.
- Specifically, this tests ONE client, sending N requests

The case we want to test is
- N clients sending 1 request

I used threading in the previous example to simulate this:
- https://github.com/robertgshaw2-neuralmagic/vllm-qa-basic-correctness/blob/main/client.py#L33

I think we should recreate this here

derekk-nm commented on neuralmagic/tests/test_basic_correctness.py:
```diff
@@ -0,0 +1,124 @@
+        api_server_args["--sparsity"] = sparsity
+
+    with VllmServer(api_server_args, logger):
+        completion = await client.completions.create(model=model,
```
@robertgshaw2-neuralmagic , are all of the models in the test "chat" models?  some say "chat" in their name, but not all.

derekk-nm replied:
@robertgshaw2-neuralmagic , I've made the necessary changes for this, but we'll need to discuss this some more. for the one model I've run through this with the responses between HuggingFace and vllm server are not very close at all when using the chat mode.  they seemed to be much closer w/ the non-chat mode of requests.

derekk-nm commented on neuralmagic/tests/test_basic_correctness.py:
```diff
@@ -0,0 +1,124 @@
+
+    with VllmServer(api_server_args, logger):
+        completion = await client.completions.create(model=model,
+                                                     prompt=example_prompts,
```
Ok, it wasn't mentioned in the [Basic Correctness Automation](https://docs.google.com/document/d/1eYWX4-YHLhfBsBvVJYM3FLmZD-QDvSOPYXYLxG3x6fI/edit?usp=sharing) doc, so I forgot to think about how your client script worked.  I'll refactor the client submissions accordingly.

derekk-nm replied:
I've added an async function to accomplish the same thing.

derekk-nm commented on tests/conftest.py:
```diff
@@ -346,6 +346,74 @@ def generate_greedy_logprobs_nm(
+            all_logprobs.append(seq_logprobs_lst)
+            seq_ids = output.sequences[0]
+            output_len = seq_ids.shape[0] - input_ids.shape[1]
+            output_ids = seq_ids[-output_len:]
```
Great!  I'm happy that this should fix the white space issue!  I'll come up with a solution that can be shared w/ your `generate_greedy_logprobs_nm()` method.

derekk-nm commented on tests/models/compare_utils.py:
```diff
@@ -1,33 +1,68 @@
+        # logprobs
+        all_outputs_df[
+            f"{name_0}_token_in_{name_1}_logprobs"] = all_outputs_df.apply(
+                lambda row: any(kee.strip() in row[f"{name_0}_tokens_index"]
```
I did not intend to keep these calls to `.strip()`.  it was an experiment to see what the results would be like w/ the leading spaces missing.

derekk-nm commented:
I believe this latest commit correctly decodes the token ids from the HFRunnerNM, but I can't find a way to decode the token ids in the logprobs keys using _decode_token_by_position_index.  The call to topk = tok_logprobs.topk() (line 416 in conftest.py) then followed by the call for topk.indices[0] (line 427) is returning the actual token_ids.  I can't seem to find any useful explanation of how topk.indices() is defined, but the doc that I found specifically states that topk will return the indices to the tensor.  I don't know which array the logprobs come from.  when I attempted to use the keys in topk.indices() as an index to seq_ids, the decoded values didn't make any sense (straight decode or passed to self._decode_token_by_position_index.  

what I've done is to tokenize the top logprob keys using self.tokenizer.decode .  This leaves us with the missing leading space for the huggingface output, so my compare util is stripping white space for the comparison.  This is not ideal, I know, but it seems to be working.  most of the prompts pass with flying colors.  Certainly open to suggestions!

With that said, there are a number of prompts that result in responses that fail the "closeness" test we're using here.  A manual review of the results (latest run output attached) demonstrates that conceptually, the responses are similar, but grammatically, they are not, so it's reasonable for them to fail this test.  @robertgshaw2-neuralmagic , we'll need to discuss which model to choose to use temporarily.  My run yesterday of all of the models in the current list of params had none pass.

This result prompted me to do additional exploration for a way to get the OpenAI server to generate token ids instead of text.  ChatGPT and Groq both seem to have hallucinated some interesting args that simply aren't real (at least not for the version of OpenAI that we're using).  I still don't think it's currently possible to get the openai server to generate output that matches for format of the HuggingFace runner.

re-reading other comments here, I know that I still have to move things around.

derekk-nm commented:
[server_basic_correctness_mistral_results_202404301351.txt](https://github.com/neuralmagic/nm-vllm/files/15170475/server_basic_correctness_mistral_results_202404301351.txt)
this attachment is the latest output from the test execution.  There are two tests using the same model, one test using a single gpu, and another with 2 (via the tensor_parallel_size param).  If you scroll to the lines that start with "E", you'll get to the error messages for the failing tests.  As you can see, the majority of prompts successfully compare between the HF and vllm server implementations, but the multi-gpu env is more "brittle" for this test.  

It's not clear to me if the issue is that the token strings in the logprobs dictionary were created by decoding the token id w/out any lookback (thus possibly missing adding a potential option in the logprobs that would have allowed the prompt to pass the test), or if there's an actual bug in the server implementation causing these prompts to generate results that do not match the HuggingFace implementation.

dbarbuzzi commented on neuralmagic/tests/test_basic_correctness.py:
```diff
@@ -0,0 +1,212 @@
+
+    asyncio_event_loop = asyncio.get_event_loop()
+    temperature = 0.0
+    with ServerContext(api_server_args, logger=logger) as _:
```
Minor/not required, but if you don’t need the server itself in the block, you can omit the `as _`:

```suggestion
    with ServerContext(api_server_args, logger=logger):
```

derekk-nm commented:
Thank you @robertgshaw2-neuralmagic for the consultation.  I've updated the script to use the existing comparison function, and corrected the method that decodes the logprob keys.  The two tests are now passing with the single model.
@dbarbuzzi , let's coordinate on the location of the tests and merging of to main.
